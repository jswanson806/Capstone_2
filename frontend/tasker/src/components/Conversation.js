import React, {useEffect, useState, useRef} from 'react';
import NewMessageForm from './NewMessageForm.js';
import { ModalBody, ModalHeader } from 'reactstrap';
import "./styles/Conversation.css";


const Conversation = ({user, currUser, messages, jobId, onMessageSent, convoId, onClose}) => {

    const [convoTitle, setConvoTitle] = useState('');

    const messagesContainerRef = useRef(null);

    useEffect(() => {
        // When messages change, scroll to the bottom of the container
        messagesContainerRef.current.scrollTop = messagesContainerRef.current.scrollHeight;
    }, [messages]);

    useEffect(() => {
        setConvoTitle(generateConvoTitle(user));
        
    }, [])


    const generateConvoTitle = (user) => {
        return `${user.firstName} ${user.lastName.slice(0, 1) + "."}`
    }
    

    return (
        <div className="convo-container">
            <div className="convo-centered-card">
            {/* renders the name of the user that is being messaged as <firstName + lastName initial + '.'> */}
            <ModalHeader>{convoTitle}</ModalHeader>
            <ModalBody style={{ maxHeight: "700px", display: "flex", flexDirection: "column-reverse", overflowY: "auto"}}>
            <div className="convo-new-message-form">
                    <NewMessageForm 
                        convoId={convoId} 
                        assignedUser={user} 
                        currUser={currUser} 
                        jobId={jobId} 
                        onMessageSent={onMessageSent}
                        onClose={onClose}
                    />
                </div>
                {/* displays the messages when messages[0] is truthy */}
                <div 
                    className="convo-messages-container"
                    ref={messagesContainerRef}
                >
                    {messages}
                </div>
                
                </ModalBody>
            </div>
        </div>
    )
}

export default Conversation;