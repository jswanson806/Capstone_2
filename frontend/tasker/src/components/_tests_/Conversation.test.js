import React from "react";
import { render } from '@testing-library/react';
import Conversation from "../Conversation.js";
import "reactstrap";

const onClose = jest.fn();
const onMessageSent = jest.fn();

const currUser = {id: 1, first_name: "Joe", last_name: "Doe", email: "test@email.com", is_worker: true, applications: [1]};
const targetUser = {id: 2, first_name: "Jane", last_name: "Doe", email: "test@email.com", is_worker: false, applications: [1]};
const jobId = 1;

describe('smoke and snapshot tests', () => {

    test('Conversation component renders correctly', async () => {

        render(
            <Conversation 
                user={targetUser}
                currUser={currUser}
                jobId={jobId}
                onMessageSent={onMessageSent}
                convoId={'u1u2j1'}
                onClose={onClose}
            />
        );
    });

    test('Conversation component matches snapshot', async () => {

        const {asFragment} = render(
            <Conversation 
                user={targetUser}
                currUser={currUser}
                jobId={jobId}
                onMessageSent={onMessageSent}
                convoId={'u1u2j1'}
                onClose={onClose}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

});

describe('useEffect hook - works', () => {

    test('sets conversation title correctly', async () => {

        const {container} = render(
            <Conversation 
                user={targetUser}
                currUser={currUser}
                jobId={jobId}
                onMessageSent={onMessageSent}
                convoId={'u1u2j1'}
                onClose={onClose}
            />
        );

        expect(container).not.toHaveTextContent("Jane Doe");
        expect(container).toHaveTextContent("Jane D.");
    });


});