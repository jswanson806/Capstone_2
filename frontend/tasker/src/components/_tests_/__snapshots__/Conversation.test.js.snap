// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests Conversation component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="convo-container"
  >
    <div
      class="convo-centered-card"
    >
      <div
        class="modal-header"
      >
        <h5
          class="modal-title"
        >
          Jane D.
        </h5>
      </div>
      <div
        class="modal-body"
        style="max-height: 700px; display: flex; flex-direction: column-reverse; overflow-y: auto;"
      >
        <div
          class="convo-new-message-form"
        >
          <div
            class="new-message-container"
          >
            <div
              class="new-message-centered-card"
            >
              <form
                class=""
              >
                <div
                  class="row form-group"
                >
                  <label
                    class=""
                    for="message"
                  >
                    New Message: 
                  </label>
                  <textarea
                    class="text-input form-control"
                    data-testid="convo-message-input"
                    id="body"
                    name="body"
                    placeholder="What's on your mind?"
                    rows="1"
                    style="resize: none; overflow-y: hidden;"
                  />
                </div>
                <button
                  class="button btn btn-info"
                  data-testid="convo-form-button"
                  type="submit"
                >
                  Send
                </button>
                <button
                  class="button btn btn-danger"
                  type="button"
                >
                  Close
                </button>
              </form>
            </div>
          </div>
        </div>
        <div
          class="convo-messages-container"
          data-testid="convo-messages-container"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
